develop/data-generation/block-models.md

```diff
@@ -0,0 +1,315 @@
+
+::: tip
+`TextureMap.all()` associates all TextureKeys with the same Identifier, no matter how many of them there are!
+:::
+
+Since we want to use the Oak Log textures, but have the ``BOTTOM``, ``TOP`` and ``SIDE`` ``TextureKey``s, we need to create a new one.
+
+@[code lang=java transcludeWith=:::datagen-model-custom:texture-map](@/reference/latest/src/client/java/com/example/docs/datagen/FabricDocsReferenceModelProvider.java)
+
+=> the ``bottom`` and ``top`` faces will use `oak_log_top.png`, the sides will use `oak_log.png`.
+
+::: warning
+All `TextureKey`s in the TextureMap **have to** match all `TextureKey`s in your parent block model! 
+:::
+
+### Custom BlockStateSupplier Method {#custom-supplier-method}
+
+The `BlockStateSupplier` contains all blockstate variants, their rotation, and other options like uvlock.
+
+@[code lang=java transcludeWith=:::datagen-model-custom:supplier](@/reference/latest/src/client/java/com/example/docs/datagen/FabricDocsReferenceModelProvider.java)
+
+First, we create a new `VariantsBlockStateSupplier` using `.create()`. Then append `.coordinate()`, in which create a new `BlockStateVariantMap`. Here, pass in all of our properties - in this case `FACING` & `SINGLE`.
+Specify which model and which transformations (uvlock, rotation) is used when using `.register()`. <br>
+For example: 
+* On the first line, the block is facing north, and is single => we use the model with no rotation.
+* On the fourth line, the block is facing west, and is single => we rotate the model on the Y axis by 270°.
+* On the sixth line, the block is facing east, but isn't single => it looks like a normal oak log => we don't have to rotate it.
+
+### Custom Datagen Method {#custom-datagen-method}
+
+The last step - creating an actual method you can call and that will generate the JSONs.
+But what are the parameters for?
+1. `BlockStateModelGenerator bsmg` is used for its variables. **Creating new ones won't work!**
+2. `Block vertSlabBlock` is the block to which we will generate the JSONs.
+3. `Block fullBlock` - is the model used when the `SINGLE` property is false = the slab block looks like a full block.
+4. `TextureMap textures` defines the actual textures the model uses. See the [Using Texture Map](#using-texture-map) chapter.
+
+@[code lang=java transcludeWith=:::datagen-model-custom:gen](@/reference/latest/src/client/java/com/example/docs/datagen/FabricDocsReferenceModelProvider.java)
+
+Let's go over line by line to really understand what is going on:
+1. Here we obtain the `Identifier` representing the vertical slab model via the `.upload()` method. We will pass in our [TextureMap](#using-texture-map) when calling this method.
+2. Here we obtain the `Identifier` of the full block - in our case of Oak Log.
+3. Here the `Consumer blockStateCollector` is called, accepting all of our blockstate variants.
+4. Here we register an item model for the vertical slab.
```

Comment:
```suggestion
First, we get the `Identifier` of the single slab model with `VERTICAL_SLAB.upload()`. Then we get the `Identifier` of the full block model with `ModelIds.getBlockModelId()`, and pass those two models into `createVerticalSlabBlockStates`. The `BlockStateSupplier` gets passed into the `blockStateCollector`, so that the JSON files are actually generated.

Also, we create a model for the vertical slab item with `BlockStateModelGenerator.registerParentedItemModel()`.
```